nat-lab/tests/test_vpn.py

```diff
@@ -564,3 +564,140 @@ async def on_stdout(stdout: str) -> None:
+                    stun_limits=ConnectionLimits(1, 1),
+                ),
+                is_meshnet=False,
+            ),
+            "10.0.254.7",
+            marks=[
+                pytest.mark.windows,
+            ],
+        ),
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.WINDOWS_VM_1,
+                adapter_type=AdapterType.WireguardGo,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.WINDOWS_VM_1,
+                    stun_limits=ConnectionLimits(1, 1),
+                ),
+                is_meshnet=False,
+            ),
+            "10.0.254.7",
+            marks=[
+                pytest.mark.windows,
+            ],
+        ),
+        pytest.param(
+            SetupParameters(
+                connection_tag=ConnectionTag.MAC_VM,
+                adapter_type=AdapterType.BoringTun,
+                connection_tracker_config=generate_connection_tracker_config(
+                    ConnectionTag.MAC_VM,
+                    stun_limits=ConnectionLimits(1, 1),
+                ),
+                is_meshnet=False,
+            ),
+            "10.0.254.7",
+            marks=pytest.mark.mac,
+        ),
+    ],
+)
+async def test_vpn_connection_private_key_change(
+    alpha_setup_params: SetupParameters,
+    public_ip: str,
+) -> None:
+    def find_cfg_by_key(
+        params: SetupParameters, key: str
+    ) -> Optional[ConnectionTrackerConfig]:
+        if params.connection_tracker_config is not None:
+            for cfg in params.connection_tracker_config:
+                if cfg.get_key() == key:
+                    return cfg
+        return None
+
+    async with AsyncExitStack() as exit_stack:
+        alpha_setup_params = copy.deepcopy(alpha_setup_params)
+
+        cfg = find_cfg_by_key(alpha_setup_params, "vpn_1")
+        assert cfg is not None
+        cfg.limits = ConnectionLimits(1, 1)
+
+        env = await exit_stack.enter_async_context(
+            setup_environment(exit_stack, [alpha_setup_params])
+        )
+
+        alpha, *_ = env.nodes
+        client_conn, *_ = [conn.connection for conn in env.connections]
+        client_alpha, *_ = env.clients
+
+        ip = await asyncio.wait_for(stun.get(client_conn, config.STUN_SERVER), 5)
+        assert ip == public_ip, f"wrong public IP before connecting to VPN {ip}"
+
+        # connect to vpn as usually
+        vpn_connection, *_ = await setup_connections(
+            exit_stack, [ConnectionTag.DOCKER_VPN_1]
+        )
+        await _connect_vpn(
+            client_conn,
+            vpn_connection.connection,
+            client_alpha,
+            alpha.ip_addresses[0],
+            config.WG_SERVER,
+        )
+
+        new_secret_key = generate_secret_key()
+        new_public_key = generate_public_key(new_secret_key)
+
+        # change public key on server
+        await vpn_connection.connection.create_process(
```

Comment:
Does this fails the test on the `exit 1` case from shell script?

Reply:
Yeah, it does